concordium-smart-contract-testing/src/types.rs

```diff
@@ -78,6 +82,90 @@ pub struct Account {
     pub balance: AccountBalance,
     /// Account policy.
     pub policy:  OwnedPolicy,
+    /// Account's public keys.
+    pub keys:    AccountAccessStructure,
+}
+
+/// A signature with account's keys.
+#[derive(Debug, Clone)]
+pub struct AccountSignatures {
+    /// It is assumed that the inner `Signature` will always be for ed25519
+    /// scheme, so will have length 64 bytes.
+    pub(crate) sigs: BTreeMap<CredentialIndex, BTreeMap<KeyIndex, Signature>>,
+}
+
+impl From<AccountSignatures> for BTreeMap<CredentialIndex, BTreeMap<KeyIndex, Signature>> {
+    fn from(value: AccountSignatures) -> Self { value.sigs }
+}
+
+impl From<BTreeMap<CredentialIndex, BTreeMap<KeyIndex, Signature>>> for AccountSignatures {
+    fn from(sigs: BTreeMap<CredentialIndex, BTreeMap<KeyIndex, Signature>>) -> Self {
+        Self { sigs }
+    }
+}
+
+impl AccountSignatures {
+    /// Return the number of signatures contained in the structure.
+    pub fn num_signatures(&self) -> u32 { self.sigs.values().map(|v| v.len() as u32).sum() }
+}
+
+impl contracts_common::Serial for AccountSignatures {
+    fn serial<W: contracts_common::Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        (self.sigs.len() as u8).serial(out)?;
+        for (k, v) in self.sigs.iter() {
+            k.serial(out)?;
+            (v.len() as u8).serial(out)?;
+            for (ki, sig) in v.iter() {
+                ki.serial(out)?;
+                // ed25519 scheme tag.
+                0u8.serial(out)?;
+                out.write_all(&sig.sig)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl contracts_common::Deserial for AccountSignatures {
+    fn deserial<R: contracts_common::Read>(source: &mut R) -> contracts_common::ParseResult<Self> {
+        use contracts_common::Get;
+        let outer_len = u8::deserial(source)?;
+        let mut last = None;
+        let mut sigs = BTreeMap::new();
+        for _ in 0..outer_len {
+            let idx = source.get()?;
+            if last >= Some(idx) {
+                return Err(contracts_common::ParseError {});
+            }
+            last = Some(idx);
+            let inner_len: u8 = source.get()?;
+            let mut inner_map = BTreeMap::new();
+            let mut x = None;
+            for _ in 0..inner_len {
+                let k = source.get()?;
+                let sig = match source.get()? {
+                    SchemeId::Ed25519 => {
+                        let mut sig = vec![0u8; ED25519_SIGNATURE_LENGTH];
+                        source.read_exact(&mut sig)?;
+                        Signature { sig }
+                    }
+                };
+
+                if let Some((old_k, old_v)) = x.take() {
+                    if k <= old_k {
+                        return Err(contracts_common::ParseError {});
+                    }
+                    inner_map.insert(old_k, old_v);
+                }
+                x = Some((k, sig));
```

Comment:
Instead of line `158` and line `163`, cannot we immediately insert this pair into the `inner_map`? Why we are waiting one for loop until we have the old value and then inserting the old values into the `inner_map`.

Reply:
This is the same code that we have for map serialization elsewhere.

It ensures that keys are not duplicated, and in strictly increasing order. Doing it this way avoids a `Copy/Clone` requirement on the keys in general. Of course this does not matter in this instance, because `k` is a u8, but I would prefer to keep this same code for map deserialization that we use everywhere else.

The alternative is to have `last: Option<u8>` variable to remember the last value inserted. But it leads to very similar code.